### PR TITLE
Add mTLS capability 

### PR DESCRIPTION
**Pull request by RichardIrons-neo4j**
This PR adds mTLS authentication capability to the driver.

### CODE REVIEW

File: Neo4j.Driver/Neo4j.Driver/Internal/Connector/TcpSocketClient.cs [hunk 1]
```diff
@@ -85,6 +86,20 @@ public void Dispose()
         }
     }
 
+    private async Task<X509Certificate2Collection> GetClientCertificates()
```
Comment (thelonelyvulpes) [inline Neo4j.Driver/Neo4j.Driver/Internal/Connector/TcpSocketClient.cs]: ValueTask

File: Neo4j.Driver/Neo4j.Driver/Public/Auth/IClientCertificateProvider.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    /// <summary>
+    /// Returns the certificate to use for new connections. Must be thread-safe. The driver will call this
+    /// method every time it makes a new connection.
+    /// </summary>
+    /// <returns>The certificate to use for new connections.</returns>
+    /// <remarks>
+    /// If a new certificate is returned, existing connections using the previous certificate will not be
+    /// affected. The new certificate will only be used for new connections.
+    /// </remarks>
+    Task<X509Certificate2> GetCertificateAsync();
```
Comment (thelonelyvulpes) [inline Neo4j.Driver/Neo4j.Driver/Public/Auth/IClientCertificateProvider.cs]: ValueTask

File: Neo4j.Driver/Neo4j.Driver/Internal/Auth/RotatingClientCertificateProvider.cs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+// limitations under the License.
+
+using System.Security.Cryptography.X509Certificates;
+using System.Threading.Tasks;
+
+namespace Neo4j.Driver.Internal.Auth;
+
+internal class RotatingClientCertificateProvider : IRotatingClientCertificateProvider
+{
+    private readonly object _lock = new();
```
Comment (thelonelyvulpes) [inline Neo4j.Driver/Neo4j.Driver/Internal/Auth/RotatingClientCertificateProvider.cs]: I think use a RWLock here or we create sequential read access

File: Neo4j.Driver/Neo4j.Driver/Internal/Connector/TcpSocketClient.cs [hunk 4]
```diff
@@ -54,9 +54,10 @@ public async Task ConnectAsync(Uri uri, CancellationToken cancellationToken = de
             try
             {
                 var sslStream = CreateSecureStream(uri);
+                var clientCertificates = await GetClientCertificates().ConfigureAwait(false);
 
                 await sslStream
-                    .AuthenticateAsClientAsync(uri.Host, null, Tls12, false)
+                    .AuthenticateAsClientAsync(uri.Host, clientCertificates, Tls12, false)
```
Comment (thelonelyvulpes) [inline Neo4j.Driver/Neo4j.Driver/Internal/Connector/TcpSocketClient.cs]: I would like, while we are here to add the ability for users to specify Tls13 (if they are using .net 5+), we don't have to just seems like a good opportunity,

File: Neo4j.Driver/Neo4j.Driver/Public/Auth/IClientCertificateProvider.cs (hunk 2 above)
Comment (RichardIrons-neo4j) [reply Neo4j.Driver/Neo4j.Driver/Public/Auth/IClientCertificateProvider.cs]: Done.

File: Neo4j.Driver/Neo4j.Driver/Internal/Connector/TcpSocketClient.cs (hunk 1 above)
Comment (RichardIrons-neo4j) [reply Neo4j.Driver/Neo4j.Driver/Internal/Connector/TcpSocketClient.cs]: done

**Comment by RichardIrons-neo4j** [comment]
[[.NET] mTLS as second factor driver implementations](https://trello.com/c/lJzgkyMA/215-net-mtls-as-second-factor-driver-implementations)

File: Neo4j.Driver/Neo4j.Driver/Internal/Auth/RotatingClientCertificateProvider.cs (hunk 3 above)
Comment (RichardIrons-neo4j) [reply Neo4j.Driver/Neo4j.Driver/Internal/Auth/RotatingClientCertificateProvider.cs]: done

**Review by AndyHeap-NeoTech — APPROVED** [review]
Looks good. Interface is clear and well explained.

**Pull request merged**